odoo/addons/base/models/res_country.py

```diff
@@ -119,12 +120,14 @@ class CountryState(models.Model):
 
     @api.model
     def _name_search(self, name, args=None, operator='ilike', limit=100, name_get_uid=None):
-        if args is None:
-            args = []
-        if self.env.context.get('country_id'):
-            args = args + [('country_id', '=', self.env.context.get('country_id'))]
-        first_state_ids = self._search([('code', '=ilike', name)] + args, limit=limit, access_rights_uid=name_get_uid)
-        search_domain = [('name', operator, name)]
-        search_domain.append(('id', 'not in', first_state_ids))
-        state_ids = first_state_ids + self._search(search_domain + args, limit=limit, access_rights_uid=name_get_uid)
-        return [(state.id, state.display_name) for state in self.browse(state_ids)]
+        args = args or []
+        if 'country_id' in self.env.context:
+            args = expression.AND([args, [('country_id', '=', self.env.context.get('country_id'))]])
+
+        if operator == 'ilike' and not (name or '').strip():
+            domain = []
+        else:
+            domain = ['|', ('code', '=ilike', name), ('name', operator, name)]
+
+        state_ids = self._search(expression.AND([domain, args]), limit=limit, access_rights_uid=name_get_uid)
```

Comment:
We probably won't display first the countries that match the country code. To check if it's an issue or not.

Reply:
Yes, if possible we should preserve the order of the original name_search.